src/Components/Components/src/RenderTree/Renderer.cs

```diff
@@ -111,13 +114,54 @@ private static IComponentActivator GetComponentActivatorOrDefault(IServiceProvid
         /// </summary>
         protected internal ElementReferenceContext? ElementReferenceContext { get; protected set; }
 
+        internal HotReloadContext HotReloadContext { get; } = new();
+
+        private async void RenderRootComponentsOnHotReload()
+        {
+            await Dispatcher.InvokeAsync(async () =>
+            {
+                HotReloadContext.HotReloading = true;
+
+                try
+                {
+                    _pendingTasks = new List<Task>();
+                    foreach (var componentState in _componentStateById.Values)
+                    {
+                        if (componentState.IsRootComponent)
+                        {
+                            var parameterView = componentState.InitialParameters is null ?
+                                ParameterView.Empty :
+                                ParameterView.FromDictionary((IDictionary<string, object?>)componentState.InitialParameters);
+                            AddToPendingTasks(componentState.Component.SetParametersAsync(parameterView));
+                        }
+                    }
+
+                    await ProcessAsynchronousWork();
+                    Debug.Assert(_pendingTasks.Count == 0);
+                }
+                finally
+                {
+                    _pendingTasks = null;
+                    HotReloadContext.HotReloading = false;
+                }
+            });
+        }
+
         /// <summary>
         /// Constructs a new component of the specified type.
         /// </summary>
         /// <param name="componentType">The type of the component to instantiate.</param>
         /// <returns>The component instance.</returns>
         protected IComponent InstantiateComponent([DynamicallyAccessedMembers(Component)] Type componentType)
-            => _componentFactory.InstantiateComponent(_serviceProvider, componentType);
+        {
+            var component = _componentFactory.InstantiateComponent(_serviceProvider, componentType);
+            if (component is IReceiveHotReloadContext receiveHotReloadContext)
```

Comment:
If there any way to tell the linker to treat this condition as always-false, that would be very cool.

Actually don't you also need to wrap this whole `if` block inside a check for `HotReloadManager.IsHotReloadEnabled`? Then it might be enough to teach the linker to treat `HotReloadManager.IsHotReloadEnabled` as always-false.